src/Data/Conduit/Tar.hs

```diff
@@ -270,3 +245,482 @@ withEntries :: MonadThrow m
+                    tarPayload 0 header tarFileInfo
+        Nothing -> return offset
+
+
+
+
+
+-- | Create a tar archive by suppying `FileInfo`, which must be immediately
+-- followed by the file content, whenever its type is `FTNormal`.
+tar :: MonadResource m =>
+       ConduitM (Either FileInfo ByteString) ByteString m FileOffset
+tar = do
+    offset <- tarFileInfo 0
+    yield terminatorBlock
+    return $ offset + fromIntegral (S.length terminatorBlock)
+
+
+-- | Create a tar archive by suppying `Header`, which must be immediately
+-- followed by the file content, whenever its type is `FTNormal`.
+tarEntries :: MonadResource m =>
+            ConduitM (Either Header ByteString) ByteString m FileOffset
+tarEntries = do
+    offset <- tarHeader 0
+    yield terminatorBlock
+    return $ offset + fromIntegral (S.length terminatorBlock)
+
+
+
+-- | Turn a stream of file paths into a stream of `FileInfo` and file
+-- content. All paths will be decended into recursively.
+filePathConduit :: MonadResource m =>
+                   ConduitM FilePath (Either FileInfo ByteString) m ()
+filePathConduit = do
+    mfp <- await
+    case mfp of
+        Just fp -> do
+            fi <- liftIO $ getFileInfo $ S8.pack fp
+            case fileType fi of
+                    FTNormal         -> do
+                        yield (Left fi)
+                        sourceFile (S8.unpack (filePath fi)) .| mapC Right
+                    FTSymbolicLink _ ->
+                        yield (Left fi)
+                    FTDirectory      -> do
+                        yield (Left fi)
+                        sourceDirectory (S8.unpack (filePath fi)) .| filePathConduit
+                    fty              -> do
+                        leftover fp
+                        throwM $ TarCreationError $ "Unsupported file type: " ++ show fty
+            filePathConduit
+        Nothing -> return ()
+
+
+-- | Recursively tar all of the files and directories.
+tarFilePath :: MonadResource m => ConduitM FilePath ByteString m FileOffset
+tarFilePath = filePathConduit .| tar
+
+
+
+createTarball :: FilePath -- ^ File name for the tarball
+              -> [FilePath] -- ^ List of files and directories to include in the tarball
+              -> IO ()
+createTarball tarfp dirs = do
+    runConduitRes $ yieldMany dirs .| void tarFilePath .| sinkFile tarfp
+
+
+writeTarball :: Handle -- ^ Handle where created tarball will be written to
+             -> [FilePath] -- ^ List of files and directories to include in the tarball
+             -> IO ()
+writeTarball tarHandle dirs = do
+    runConduitRes $ yieldMany dirs .| void tarFilePath .| sinkHandle tarHandle
+
+
+pathSeparatorS :: ByteString
+pathSeparatorS = S8.singleton pathSeparator
+
+
+fileInfoFromHeader :: Header -> FileInfo
+fileInfoFromHeader header@(Header {..}) =
+    FileInfo
+    { filePath = if SS.length headerFileNamePrefix > 0
+                 then fromShort headerFileNamePrefix <> pathSeparatorS <>
```

Comment:
I'm not certain, but I _think_ using `S.concat` will be more efficient here as it will only create one output buffer and copy once.